edk2toollib/uefi/telem/parser_main.py

```diff
@@ -0,0 +1,660 @@
+
+        temp = self.SectionSeverityParse()
+        string = concat_n(string,concat("Section Severity:  ",temp))
+
+        temp = self.SectionTypeParse()
+        string = concat_n(string,concat("Section Type:      ",temp))
+
+        temp = self.FRUIDParse()
+        string = concat_n(string,concat("FRU ID:            ",temp))
+        
+        temp = self.FRUStringParse()
+        string = concat_n(string,concat("FRU String:        ",temp))
+        
+        return string
+
+##
+# Dumps byte code of input
+##
+def HexDump(input, bytesperline):
+
+    string      = "" # Stores the entire hexdump string
+    asc         = "" # Stores the ascii version of the current hexdump line
+    byte        = "" # Stores the base 16 byte version of the current hexdump line
+    rangecheck  = lambda x : x < 31 or x > 127 # Used to check if a byte value is within relevant ascii character bounds
+
+    # Go through every byte from the input
+    for i in range(len(input)):
+
+        # Add a byte string version of the byte onto the byte string
+        byte = concat_s(byte, format(input[i],'02X'))
+        
+        # Add an ascii version of the byte onto the ascii string
+        if rangecheck(input[i]):
+                asc = concat_s(asc,". ")
+        else:
+            asc = concat_s(asc,format(chr(input[i])," <2"))
+        
+        # Once we've reached bytesperline length, concatenate asc and byte strings and start a new line
+        if(not (i + 1) % bytesperline):
+            string = concat_n(string,concat(byte,asc))
+            asc = ""
+            byte = ""
+    
+    # Check if there are any remaining characters in the asc and byte strings to be added to string 
+    if(len(input) % bytesperline):  
+        string = concat(string,concat_n(concat(byte,"   " * (bytesperline - (len(byte)//3))),concat(asc,"   " * (bytesperline - (len(asc)//3)))))
+
+    return string
+
+##
+# Load friendly names from friendlynames.csv to FriendlyNames dict
+##
+def ImportFriendlyNames():
+    # open friendlynames.csv
+    with open('friendlynames.csv','r') as csv_file:
+
+        rowcounter = 2 # Used to track which row we are reading
+        csv_reader = csv.reader(csv_file, delimiter=',') # Read the csv file
+        next(csv_reader) # Skip the header
+
+        for row in csv_reader:
+            try:
+                # Input friendly name into dict
+                FriendlyNames[uuid.UUID(row[1].strip())] = row[0]
+            except:
+                # Alert user if a row could not be parsed
+                print("Unable to add row " + str(rowcounter) + " to the friendly name dictionary!")
+            
+            rowcounter += 1
+
+##
+# Load all plugins from the /plugins folder
+##
+def LoadPlugins():
+    # Get all subclasses of SECTION_PARSER_PLUGIN which have been imported 
+    # using the "from plugins import *" call at start of file
+    subclasslist = SECTION_PARSER_PLUGIN.__subclasses__()
```

Comment:
Are we running this in the context of a UEFI build? Or standalone? Perhaps it would make more sense to use the SDE plugin system? Or maybe even use both?

Reply:
Where can I find info about the SDE plugin system?